### PR TITLE
ARROW-3959: [Rust] Add date/time data types

### PR DESCRIPTION
**Pull request by andygrove**
This only adds the date/time types to the DataTypes enum as well as JSON serialization for meta data.

This PR also implements `Schema::to_json`

### CODE REVIEW

**Comment by andygrove** [comment]
Appveyor is failing due to unrelated issue ARROW-4185.

@nevi-me @sunchao @kszucs @paddyhoran looking for review pls

**Comment by nevi-me** [comment]
Looks great @andygrove, I have few questions:

The specification allows for time types to contain a timezone, how easy would it be to add this to Rust in future?

With the data types now present, what other work will we need, potentially as separate JIRAs?

I think:

* Creating types ala `impl ArrowNativeType`
* Various time arrays and their builders

**Comment by andygrove** [comment]
@nevi-me yes I think that is accurate

**Comment by andygrove** [comment]
I don't see any mention of timezone in https://github.com/apache/arrow/blob/master/docs/source/format/Metadata.rst

**Comment by nevi-me** [comment]
Sorry Andy, it's mentioned in theL

* flatbuffer schema (https://github.com/apache/arrow/blob/master/format/Schema.fbs#L182), 
* Java API (http://arrow.apache.org/docs/java/org/apache/arrow/vector/types/pojo/ArrowType.Timestamp.html#Timestamp-org.apache.arrow.vector.types.TimeUnit-java.lang.String-), 
* cpp (http://arrow.apache.org/docs/cpp/classarrow_1_1_timestamp_type.html#ad0be5cd4daf44dbbcecacfcdd6499a76).

Don't know if the omission from the spec is intentional, but I had gone on the memory that a timezone was a thing since wayback when I started following the Arrow project.

**Comment by andygrove** [comment]
Good point. I filed https://issues.apache.org/jira/browse/ARROW-4194 for this.

**Comment by nevi-me** [comment]
Do you think it's better to add timezone as part of this PR, or later? Naively, I assume it's `Timestamp(TimestampUnit, String)`, and then getting + setting it in the `from(json)` and `to_json()` methods?

It looks like from the discussions in [[ARROW-637]](https://issues.apache.org/jira/browse/ARROW-637), tz is mainly relevant when displaying to users.

**Comment by andygrove** [comment]
I would prefer to do this as separate PR after having a PR approved to update Metadata.rst, just in case this is intentionally not there.

**Comment by agrove-rms** [comment]
@nevi-me @sunchao @kszucs @paddyhoran I'd really like to get this merged in time for the 0.12 release since I'm currently blocked from supporting date/time types in DataFusion .. any chance of a review?

**Review by nevi-me — APPROVED** [review]

File: rust/arrow/src/datatypes.rs [hunk 1]
```diff
@@ -56,11 +56,44 @@ pub enum DataType {
...
+    Interval(IntervalUnit),
     Utf8,
     List(Box<DataType>),
     Struct(Vec<Field>),
 }
 
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub enum TimestampUnit {
+    Second,
+    Millisecond,
```
Comment (sunchao) [inline rust/arrow/src/datatypes.rs]: Why we need a `TimestampUnit` given we already have `TimeUnit`? they are exactly the same.

Comment (andygrove) [reply rust/arrow/src/datatypes.rs]: Good point. I checked the C++ code and they just have TimeUnit so I removed TimestampUnit now.

**Review by sunchao — APPROVED** [review]
LGTM. I think this is just some placeholder for now, is that correct? I'd expect more like array, array builders, etc., to make it fully usable.

Also it'll be great if we can address the timezone issue soon.

**Comment by agrove-rms** [comment]
Thanks @sunchao and yes, that's correct. This allows a developer to have a dependency on Arrow and define schemas using Arrow types, even though Arrow does not (yet) provide an implementation of these types.

**Pull request closed without merging**